### ISSUE TITLE
Missing framework - Beego

### ISSUE THREAD

**Issue by bharrat** [sqreen/go-agent#105]
Please add support for Beego Framework

https://github.com/astaxie/beego/

**Comment by Julio-Guerra** [comment]
Hello @bharrat,

Thank you for your interest in Sqreen for Go and letting us know you use Beego.

I just had a look at Beego and it is actually compatible with a standard `net/http` middleware, by passing it to [App.Run()](https://godoc.org/github.com/astaxie/beego#App.Run) or [RunWithMiddleWares() ](https://godoc.org/github.com/astaxie/beego#RunWithMiddleWares).

Here is a code snippet example that should do the job:

```go
package main

import (
 "github.com/sqreen/go-agent/sdk/middleware/sqhttp"
  "github.com/astaxie/beego"
)

func main() {
  app := beego.NewApp()
  // ...
  app.Run(sqhttp.Middleware)
} 
```

Let me know this doesn't work. I will add it to our doc for the next release to come.

**Comment by bharrat** [comment]
Hi @Julio-Guerra  thanks for that suggestion, I did try what  you suggested ( using `RunWithMiddlewares` option), but seems like it doesn't work, Sqreen can not intercept/monitor any requests.

**Comment by bharrat** [comment]
hi @Julio-Guerra  thanks for your suggestions. 
I tried both the options:
1. App.Run
2. and RunWithMiddlewares
and couldn't get either option to work

1. Using `App.run` basically messe up our controller registration so we can't really use that
2. with `RunWithMiddlewares`, i can see Middleware gets registered but its not  able to intercept any of the requests, seems to me something related to how Beego has implemented support for Middlewares.

FWIW, all the other middlewares we have implemented in Beego were implemented as "Filters" 

https://beego.me/docs/mvc/controller/filter.md


https://beego.me/docs/contributed/

**Comment by Julio-Guerra** [comment]
Ok. Thanks for the pointers, that's indeed what I was actually looking for.
I should be able to add it for v0.10, likely for the end of next week.

**Comment by Julio-Guerra** [comment]
@bharrat I am currently working on beego's support and so far the standard `net/http` middleware solution is the one working the best.

You were saying it wasn't working - can you tell me more about the issue?

**Comment by bharrat** [comment]
Thanks for following up @Julio-Guerra . So issue I was getting with our app was middleware code was not getting executed. After your comment I created a sample app and found the cause of why middleware is not working, looks like its a bug on Beeego side, when this config is set:
`beego.BConfig.Listen.Graceful = true`

Our app is setting this flag and this is the reason why middleware was not working :)


See sample code below:
```
package main

import (
	"fmt"
	"net/http"

	"github.com/astaxie/beego"
)
type MainController struct {
	beego.Controller
}

func (m *MainController) Get() {
	m.Ctx.WriteString(fmt.Sprintf("hello world"))
}

func main() {
	
	// This is the cause of middleware not working
	beego.BConfig.Listen.Graceful = true
	
	beego.Router("/", &MainController{})
	beego.RunWithMiddleWares(":8084", func(next http.Handler) http.Handler {
		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			fmt.Printf("log: in  the middleware from: %s\n", r.RequestURI)
			next.ServeHTTP(w, r)
		})
	})
}

```

**Comment by Julio-Guerra** [comment]
Indeed, this changes the server start conditions which doesn't use the HTTP handler value having the middleware applied.

I was about to create a PR for Beego to fix the issue but I just saw yours was already waiting https://github.com/astaxie/beego/pull/3996 :smile:

Unfortunately, it looks like Beego isn't seriously maintained, so I fear that you'll have to wait for a very long time before it gets applied... :confused:

Anyway, this issue allowed us to improve our coverage of standard `net/http` handlers that you would still get into v0.10. The PR is coming and will close this issue.

Thanks again for your technical help on this :+1:

**Comment by Julio-Guerra** [comment]
Closed by #110
